clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -2234,8 +2232,29 @@ static std::string getKernelNameTypeString(QualType T, ASTContext &Ctx,
 
     return eraseAnonNamespace(ArgOS.str().str());
   }
-
-  return eraseAnonNamespace(FullyQualifiedType.getAsString(TypePolicy));
+  if (const auto *D = dyn_cast<CXXRecordDecl>(RD)) {
+    std::string NSStr = "";
+    const DeclContext *DC = D->getDeclContext();
+    SmallString<64> Buf;
+    llvm::raw_svector_ostream OS(Buf);
+    while (DC) {
+      auto *NS = dyn_cast<NamespaceDecl>(DC);
+      if (!NS || (NS && NS->getName().empty())) {
+        if (DC->isTranslationUnit())
+          NSStr.append(Twine("::" + D->getName()).str());
```

Comment:
The Twine function-like-cast is really odd here.

Reply:
Proposed a new solution to this.